-1

Tested at: 559a8462965e3986960a9e91d510e6955ed65844

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23480/28472/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found an error when building: 
 
 <pre>tmp/slc6_amd64_gcc630/src/SimG4CMS/Muon/src/SimG4CMSMuon/MuonGEMFrameRotation.cc.o:(.data.rel.ro._ZTI20MuonGEMFrameRotation[_ZTI20MuonGEMFrameRotation]+0x10): undefined reference to `typeinfo for MuonFrameRotation'
tmp/slc6_amd64_gcc630/src/SimG4CMS/Muon/src/SimG4CMSMuon/MuonSensitiveDetector.cc.o: In function `MuonSensitiveDetector::MuonSensitiveDetector(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DDCompactView const&, SensitiveDetectorCatalog const&, edm::ParameterSet const&, SimTrackManager const*)':
MuonSensitiveDetector.cc:(.text+0x29cd): undefined reference to `vtable for MuonFrameRotation'
tmp/slc6_amd64_gcc630/src/SimG4CMS/Muon/src/SimG4CMSMuon/MuonEndcapFrameRotation.cc.o:(.data.rel.ro._ZTI23MuonEndcapFrameRotation[_ZTI23MuonEndcapFrameRotation]+0x10): undefined reference to `typeinfo for MuonFrameRotation'
tmp/slc6_amd64_gcc630/src/SimG4CMS/Muon/src/SimG4CMSMuon/MuonME0FrameRotation.cc.o:(.data.rel.ro._ZTI20MuonME0FrameRotation[_ZTI20MuonME0FrameRotation]+0x10): undefined reference to `typeinfo for MuonFrameRotation'
collect2: error: ld returned 1 exit status
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-04-2300/src/SimG4CMS/ShowerLibraryProducer/plugins/module.cc 
gmake: *** [tmp/slc6_amd64_gcc630/src/SimG4CMS/Muon/src/SimG4CMSMuon/libSimG4CMSMuon.so] Error 1
Leaving library rule at SimG4CMS/Muon
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_2_X_2018-06-04-2300/src/SimG4CMS/ShowerLibraryProducer/plugins/HcalForwardLibWriter.cc 
>> Building big object file tmp/slc6_amd64_gcc630/src/SimG4CMS/Forward/plugins/SimG4CMSForwardPlugins/bigobj/SimG4CMSForwardPlugins.obj
</pre>

--- a/SimG4CMS/Muon/src/MuonFrameRotation.cc
+++ b/SimG4CMS/Muon/src/MuonFrameRotation.cc
@@ -1,0 +1,6 @@
+#include "SimG4CMS/Muon/interface/MuonFrameRotation.h"
+#include "G4Step.hh"
+
+Local3DPoint MuonFrameRotation::transformPoint(const Local3DPoint& point,const G4Step*) const {
+  return point;
+} 

--- a/SimG4CMS/Muon/src/MuonSensitiveDetector.cc
+++ b/SimG4CMS/Muon/src/MuonSensitiveDetector.cc
@@ -41,7 +41,7 @@ MuonSensitiveDetector::MuonSensitiveDetector(const std::string& name,
     thePV(nullptr), theHit(nullptr), theDetUnitId(0), theTrackID(0), theManager(manager)
 {
   edm::ParameterSet m_MuonSD = p.getParameter<edm::ParameterSet>("MuonSD");
-  ePersistentCutGeV = m_MuonSD.getParameter<double>("EnergyThresholdForPersistency");//Default 1. GeV
+  ePersistentCutGeV = m_MuonSD.getParameter<double>("EnergyThresholdForPersistency")/CLHEP::GeV;//Default 1. GeV
   allMuonsPersistent = m_MuonSD.getParameter<bool>("AllMuonsPersistent");
   printHits  = m_MuonSD.getParameter<bool>("PrintHits");
   
@@ -125,6 +125,9 @@ bool MuonSensitiveDetector::ProcessHits(G4Step * aStep, G4TouchableHistory * ROh
       } else {
 	updateHit(aStep);
       } 
+    } else {
+      thePV = aStep->GetPreStepPoint()->GetPhysicalVolume();
+      theTrackID = aStep->GetTrack()->GetTrackID();
     }
   }
   return true;
@@ -154,8 +157,7 @@ uint32_t MuonSensitiveDetector::setDetUnitId(const G4Step * aStep)
 bool MuonSensitiveDetector::newHit(const G4Step * aStep){
 
   return (!theHit || (aStep->GetTrack()->GetTrackID() != theTrackID) 
-	  || (aStep->GetPreStepPoint()->GetPhysicalVolume() != thePV) 
-	  || (setDetUnitId(aStep) !=  theDetUnitId));
+	  || (aStep->GetPreStepPoint()->GetPhysicalVolume() != thePV));
 }
 
 void MuonSensitiveDetector::createHit(const G4Step * aStep){

--- a/SimG4CMS/Tracker/interface/TkAccumulatingSensitiveDetector.h
+++ b/SimG4CMS/Tracker/interface/TkAccumulatingSensitiveDetector.h
@@ -51,11 +51,12 @@ private:
     bool newHit   (const G4Step *);
     bool closeHit (const G4Step *);
 
+protected:
     void update(const BeginOfEvent *) override;
     void update(const BeginOfTrack *) override;
     void update(const BeginOfJob *) override;
 
-    Local3DPoint RotateAndScale(const Local3DPoint&);
+private:
 
     // data members initialised before run
     const SimTrackManager* theManager;

--- a/SimG4CMS/Tracker/src/TkAccumulatingSensitiveDetector.cc
+++ b/SimG4CMS/Tracker/src/TkAccumulatingSensitiveDetector.cc
@@ -308,13 +308,13 @@ void TkAccumulatingSensitiveDetector::updateHit(const G4Step * aStep)
     mySimHit->addEnergyLoss(theEnergyLoss);
     if (printHits) {
       globalExitPoint = ConvertToLocal3DPoint(aStep->GetPostStepPoint()->GetPosition());
-    }
-    LogDebug("TrackerSimDebug")
+      LogDebug("TrackerSimDebug")
       << " updateHit: for " << aStep->GetTrack()->GetDefinition()->GetParticleName()
       << " trackID= " << aStep->GetTrack()->GetTrackID() << " deltaEloss= " << theEnergyLoss
       << "\n Updated PSimHit: " << mySimHit->detUnitId() << " " << mySimHit->trackId()
       << " " << mySimHit->energyLoss() << " " << mySimHit->entryPoint() 
       << " " << mySimHit->exitPoint();
+    }
 }
 
 bool TkAccumulatingSensitiveDetector::newHit(const G4Step * aStep)

